src/app/api/quality_matrix/collections.py

```diff
@@ -0,0 +1,65 @@
+from uuid import UUID
+
+from elasticsearch_dsl import Q
+from elasticsearch_dsl.response import Response
+
+from app.api.quality_matrix.quality_matrix import (
+    QUALITY_MATRIX_RETURN_TYPE,
+    _quality_matrix,
+    add_base_match_filters,
+    extract_sources_from_response,
+    get_properties,
+)
+from app.core.config import ELASTIC_TOTAL_SIZE
+from app.core.constants import PORTAL_ROOT_ID
+from app.elastic.dsl import qbool
+from app.elastic.search import Search
+
+_TITLE_PROPERTY = "properties.cm:title"
+
+
+def all_collections(node_id: UUID = PORTAL_ROOT_ID) -> dict[str, int]:
+    aggregation_name = "uniquefields"
+    s = add_base_match_filters(
+        Search()
+        .query(
+            qbool(
+                must=[
+                    Q("term", **{"path": node_id}),
+                ]
+            )
+        )
+        .source(includes=["aggregations", _TITLE_PROPERTY])
+    )
+    s.aggs.bucket(aggregation_name, "terms", field="path", size=ELASTIC_TOTAL_SIZE)
+
+    response: Response = s.execute()
+    return extract_sources_from_response(response, aggregation_name)
+
+
+async def id_to_title_mapping(node_id: UUID):
```

Comment:
Having another look at that, I would use some "builder pattern" for the whole elastic search query setup.
As we already have the `Search` class, why not add methods such as 

```python
class Search(...):
     ...
    def add_base_match_filters(self)->'Search':
         """Setup some global filters.."""
   def collection_counts(self, args)->'Search':
         """Add bucket aggregation to get the counts per collection"""
```

Then one can build and run the actual searches as 

```python
data = (
        Search()
               .add_base_match_filters()
               .collection_counts(args="some value")
               .execute()
          )
```

Imho that is much easier to understand and maintain :P

Reply:
These are some interesting ideas - I wanted to refactor the Search class soon anyway.
In the current state I vote against handing the Search object into these functions, in the current state it feels - unnecessary. I do see your point indicating "intention hiding", i.e., the IO logic is hidden behind titles, which just tell what is handed back, not where it is coming from. I will create an issue, since refactoring the elasticsearch stuff - to e.g. get rid of technical debt - is a useful thing to do, anyway.

Reply:
Added to: https://github.com/openeduhub/metaqs-main/issues/45